pkg/kv/kvserver/apply/asserter.go

```diff
@@ -0,0 +1,145 @@
+	appliedCmds map[roachpb.RangeID]map[kvserverbase.CmdIDKey]*appliedCmd
+}
+
+// NewAsserter creates a new asserter.
+func NewAsserter() *Asserter {
+	return &Asserter{
+		appliedIndex: map[roachpb.RangeID]map[roachpb.ReplicaID]kvpb.RaftIndex{},
+		appliedCmds:  map[roachpb.RangeID]map[kvserverbase.CmdIDKey]*appliedCmd{},
+	}
+}
+
+// ensureRangeLocked ensures that data structures exist for the given range.
+// Asserter.mu must be held.
+func (a *Asserter) ensureRangeLocked(rangeID roachpb.RangeID) {
+	if _, ok := a.appliedCmds[rangeID]; !ok {
+		a.appliedCmds[rangeID] = map[kvserverbase.CmdIDKey]*appliedCmd{}
+	}
+	if _, ok := a.appliedIndex[rangeID]; !ok {
+		a.appliedIndex[rangeID] = map[roachpb.ReplicaID]kvpb.RaftIndex{}
+	}
+}
+
+// Apply records a command being applied, and asserts validity.
+//
+// It assumes that entry application is durable, i.e. that nodes won't lose
+// applied state. If needed, we can loosen this requirement by signalling
+// truncation of applied state (e.g. on node restart).
+func (a *Asserter) Apply(
+	rangeID roachpb.RangeID,
+	replicaID roachpb.ReplicaID,
+	cmdID kvserverbase.CmdIDKey,
+	entry raftpb.Entry,
+) {
+	if len(cmdID) == 0 {
+		return
+	}
+
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	a.ensureRangeLocked(rangeID)
+
+	// Assert and record the applied index.
+	if appliedIndex := a.appliedIndex[rangeID][replicaID]; kvpb.RaftIndex(entry.Index) <= appliedIndex {
+		panic(fmt.Sprintf("applied index regression for r%d/%d: %d -> %d", rangeID, replicaID, appliedIndex, entry.Index))
+	}
+	a.appliedIndex[rangeID][replicaID] = kvpb.RaftIndex(entry.Index)
+
+	if ac, ok := a.appliedCmds[rangeID][cmdID]; !ok {
+		// New command was applied, record it.
+		a.appliedCmds[rangeID][cmdID] = &appliedCmd{
+			cmdID:    cmdID,
+			rangeID:  rangeID,
+			logIndex: kvpb.RaftIndex(entry.Index),
+			replicas: map[roachpb.ReplicaID]bool{replicaID: true},
+		}
+	} else if ac.logIndex == kvpb.RaftIndex(entry.Index) {
+		// Command applying at the expected index, record the replica.
+		ac.replicas[replicaID] = true
+	} else {
+		// Applied command at unexpected log index, bail out.
+		var replicas []roachpb.ReplicaID
+		for id := range ac.replicas {
+			replicas = append(replicas, id)
+		}
+		msg := fmt.Sprintf("command %s re-applied at index %d on r%d/%d\n", cmdID, entry.Index, rangeID, replicaID)
+		msg += fmt.Sprintf("previously applied at index %d on replicas %s\n", ac.logIndex, replicas)
+		msg += fmt.Sprintf("entry: %+v", entry) // TODO: this formatting is not helpful
+		panic(msg)
+	}
+}
+
+// ApplySnapshot records snapshot application at the given applied index. The
+// replica is considered to have applied all commands up to and including the
+// applied index.
+func (a *Asserter) ApplySnapshot(
+	rangeID roachpb.RangeID,
+	replicaID roachpb.ReplicaID,
+	index kvpb.RaftIndex,
+) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	a.ensureRangeLocked(rangeID)
+
+	// Record the applied index.
+	a.appliedIndex[rangeID][replicaID] = index
```

Comment:
What we could/should verify for a snapshot is:
- There is at least one applied command at the same index (i.e. someone has applied a command, and then sent a snapshot including it).
- (exception: the initial snapshot index, I think it's 10 or something; need to dig it)
- The LAI (and maybe some other metadata like CT) of this command matches what the snapshot state says.

Reply:
I'll leave this for later -- we need to start getting some runtime in the nightlies to look for reapplication.